README
NAME
    App::mojopaste - Pastebin application

VERSION
    0.25

DESCRIPTION
    Mojopaste is a pastebin application. There's about one million of these
    out there, but if you have the need to run something internally at work
    or you just fancy having your own pastebin, this is your application.

  Text and code
    The standard version of App::mojopaste can take normal text as input,
    store it as a text file on the server and render the content as either
    plain text or prettified using Google prettify
    <https://code.google.com/p/google-code-prettify/>. (Note: Maybe another
    prettifier will be used in future versions)

  Charts
    In addition to just supporting text, this application can also make
    charts from the input data. To turn this feature on, you need to specify
    "enable_charts" in the config or set the "PASTE_ENABLE_CHARTS"
    environment variable:

      $ PASTE_ENABLE_CHARTS=1 script/mojopaste daemon;

    The input chart data must be valid CSV:

    CSV data is similar to "Just data" above, except the first line is used
    as "xkey,ykey1,ykey2,...". Example:

      # Can have comments in CSV input as well
      x,a,b
      2015-02-04 15:03,120,90
      2015-03-14,75,65
      2015-04,100,40

    CSV input data require Text::CSV to be installed.

  Embedding
    A paste can be embedded in other pages using the query param "embed".
    Examples:

    * <http://p.thorsen.pm/mojopastedemo.txt>

      Get the raw data.

    * <http://p.thorsen.pm/mojopastedemo?embed=text>

      Show the paste without any margin/padding and no menu.

    * <http://p.thorsen.pm/mojopastedemo/chart?embed=graph>

      Show only the graph data.

    * <http://p.thorsen.pm/mojopastedemo/chart?embed=graph,heading,descripti
      on>

      Show the graph data, heading and description, but no menus.

DEMO
    You can try mojopaste here: <http://p.thorsen.pm>.

INSTALLATION
    Install system wide with cpanm:

      $ cpanm --sudo App::mojopaste

    Don't have cpanm installed?

      $ curl -L http://cpanmin.us | perl - --sudo App::mojopaste
      $ wget http://cpanmin.us -O - | perl - --sudo App::mojopaste

SYNOPSIS
    * Simple single process daemon

        $ mojopaste daemon --listen http://*:8080

    * Save paste to custom dir

        $ PASTE_DIR=/path/to/paste/dir mojopaste daemon --listen http://*:8080

    * Using the UNIX optimized, preforking hypnotoad web server

        $ MOJO_CONFIG=/path/to/mojopaste.conf hypnotoad $(which mojopaste)

      Example mojopaste.conf:

        {
          paste_dir     => '/path/to/paste/dir',
          enable_charts => 1, # default is 0
          hypnotoad => {
            listen => ['http://*:8080'],
          },
        }

      "enable_charts" is for adding a button which can make a chart of the
      input data using morris.js <http://morrisjs.github.io/morris.js>

      Check out Mojo::Server::Hypnotoad for more hypnotoad options.

OTHER PASTEBINS
    * <http://paste.scsys.co.uk>

    * <http://paste.ubuntu.com>

    * <http://pastebin.com>

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution App-mojopaste

1.06 Not Released
 - Add support for X-Request-Base

1.05 2021-06-17T07:56:14+0900
 - Compatible with Mojolicious 9.11
   Contributor: Dan Book

1.04 2019-08-09T16:37:47+0200
 - Fix failing tests #19

1.03 2019-08-08T23:51:56+0200
 - Add support for custom branding

1.02 2019-08-08T21:47:21+0200
 - Fix height of textarea

1.01 2019-08-08T21:25:41+0200
 - Changed layout to "dark mode"
 - Changed layout to using "modern" CSS technologies, such as var()
 - Swapped Google Code Prettify style to
   https://jmblog.github.io/color-themes-for-google-code-prettify/themes/atelier-dune-dark.min.css
 - Using a modified version of https://commons.wikimedia.org/wiki/File:Gnome-edit-paste.svg
   as logo, under the CC BY-SA 3.0 license.
 - Add basic documentation to script #17
   Contributor: Dan Book
 - Fix shebang #16
   Contributor: Dan Book

1.00 2018-09-26T00:00:33+0900
 - Compatible with Mojolicious 8.00
 - Changed from delay to promises

0.26 2018-04-21T10:37:22+0200
 - Add title that describe which paste you look at
 - Add support for pluggable backends #14
 - Change Dockerfile to pull from github.com/jhthorsen/app-mojopaste instead of cpan
 - Fix "...did you mean do "./script/mojopaste"? at t/config.t line 10."

0.25 2017-06-20T00:20:46+0200
 - Mojolicious has deprecated Mojo::Util::slurp() and spurt()
 - Fix failing tests

0.24 2016-10-13T10:01:53+0200
 - Writing/reading files are more robust

0.23 2016-09-08T20:08:22+0200
 - Fix rendering 404 for paste_not_found() and .txt extension

0.22 2016-09-06T21:18:39+0200
 - Will never allow robots
 - Will only graph csv data

0.21 2016-06-30T14:44:51+0200
 - Change default paste directory
 - Can embed paste and graphs
 - Fix using the whole window and nothing more when showing a graph

0.20 2016-06-29T18:08:30+0200
 - Fix X-Plain-Text-URL must be an absolute URL
 - Add line numbers to pretty printed code
 - Change "Paste" button to "Save"

0.19 2016-06-29T13:59:26+0200
 - Fix selecting only pastebin text on ctrl+a or cmd+a
 - Fix handling of empty paste
 - Add EXPERIMENTAL X-Plain-Text-URL header to paste web page
 - Add support for header and description for graph view #7
 - Change from dark to light color theme

0.18 2016-01-15T07:52:58+0100
 - Try to fix failing MSWin32 test
   http://www.cpantesters.org/cpan/report/873ffedb-6bf4-1014-a355-7b0739136c96

0.17 2016-01-13T21:36:27+0100
 - Fix unicode.t by bumping Mojolicious version #10
 - Cannot make empty paste
 - Non-existing paste will result in 404 instead of 500 #8
   Contributor: Stephan Jauernick
 - Use File::Spec to construct correct paths on all OS #8
   Contributor: Stephan Jauernick

0.16 2016-01-13T09:25:51+0100
 - Recommends JSON::Syck
 - Recommends Text::CSV
 - No need to encode/decode UTF-8 anymore

0.15 2015-04-30T10:17:13+0200
 - Change "Powered by" URL
 - Change styling for .morris-hover
 - Change 404 page to show that the paste could not be found.

0.14 2015-04-20T17:03:36+0200
 - Fix unit tests

0.13 2015-04-20T16:54:30+0200
 - SECURITY FIX! Cannot read ../../private/file #6
 - Capture chart errors
 - Will construct default ykeys from all chart elements

0.11 2015-04-20T15:15:43+0200
 - Can make charts using morris.js

0.10 2014-08-07T17:51:23Z
 - Fix button border styling, #2

0.09 2014-08-07T11:14:00Z
 - Add support for .txt as raw format
 - Add focus on textarea on load
 - Change to plain javascript (no jQuery)
 - New design with bottom navbar

0.08 2014-07-22T17:00:15Z
 - Compatible with Mojolicious 5.xx

0.07 2014-07-21T06:52:09Z
 - Compatible with Mojolicious 5.xx

0.06 2014-07-18T02:13:05Z
 - Change bugracker

0.0501 2014-05-30T08:02:31Z
 - Fix fragile test t/mojopaste.t

0.05 2013-12-31T19:39:07Z
 - Allow pasting without javascript enabled

0.0403 2013-09-09T16:27:19Z
 - Fix showing unicode characters

0.0402 2013-09-09T16:08:21Z
 - Fix pasting unicode characters

0.0401 2013-09-06T23:34:51Z
 - Fix Makefile.PL

0.04 2013-09-05T23:55:30Z
 - Fix including jquery.js from Mojo
 - Render as HTML5
 - Looks better on iPhone
 - Render 404 when paste is not found
 - Trying to trick the dumbest robots

0.03 2013-09-05T16:54:59Z
 - Fix typo in PREREQ_PM

0.02 2013-09-05T16:48:08Z
 - Fix repository path

0.01 2013-09-05T16:46:19Z
 - Add bin/mojopaste
 - Add documentation



t/Helper.pm
package t::Helper;
use Mojo::Base -strict;
use Cwd ();
use File::Spec;
use Test::More ();

sub t {
  Test::More::plan(skip_all => $@) unless do File::Spec->catfile(qw(script mojopaste));
  Test::More::plan(skip_all => "$ENV{PASTE_DIR} was not created") unless -d $ENV{PASTE_DIR};
  return Test::Mojo->new;
}

sub import {
  my $caller = caller;

  $_->import for qw(strict warnings utf8);
  $ENV{PASTE_DIR} = File::Spec->catdir(qw(t paste));

  eval <<"HERE" or die $@;
package $caller;
use Test::Mojo;
use Test::More;
1;
HERE
}
1;


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use utf8;
use ExtUtils::MakeMaker;
my %WriteMakefileArgs = (
  NAME           => 'App::mojopaste',
  AUTHOR         => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE        => 'artistic_2',
  ABSTRACT_FROM  => 'lib/App/mojopaste.pm',
  VERSION_FROM   => 'lib/App/mojopaste.pm',
  EXE_FILES      => [qw(script/mojopaste)],
  OBJECT         => '',
  BUILD_REQUIRES => {},
  TEST_REQUIRES  => {'Test::More'  => '0.88'},
  PREREQ_PM      => {'Mojolicious' => '9.11', 'Text::CSV' => '1.30'},
  META_MERGE     => {
    'dynamic_config' => 0,
    'meta-spec'      => {version => 2},
    'resources'      => {
      bugtracker => {web => 'https://github.com/jhthorsen/app-mojopaste/issues'},
      homepage   => 'https://github.com/jhthorsen/app-mojopaste',
      repository => {
        type => 'git',
        url  => 'https://github.com/jhthorsen/app-mojopaste.git',
        web  => 'https://github.com/jhthorsen/app-mojopaste',
      },
    },
    'x_contributors' => ['Jan Henning Thorsen <jhthorsen@cpan.org>'],
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/App/mojopaste.pm
package App::mojopaste;

our $VERSION = '1.05';

1;

=encoding utf8

=head1 NAME

App::mojopaste - Pastebin application

=head1 VERSION

1.05

=head1 DESCRIPTION

Mojopaste is a pastebin application. There's about one million of these out
there, but if you have the need to run something internally at work or you
just fancy having your own pastebin, this is your application.

=head2 Text and code

The standard version of L<App::mojopaste> can take normal text as input,
store it as a text file on the server and render the content as either
plain text or prettified using L<Google prettify|https://code.google.com/p/google-code-prettify/>.
(Note: Maybe another prettifier will be used in future versions)

=head2 Charts

In addition to just supporting text, this application can also make charts
from the input data. To turn this feature on, you need to specify
"enable_charts" in the config or set the C<PASTE_ENABLE_CHARTS>
environment variable:

  $ PASTE_ENABLE_CHARTS=1 script/mojopaste daemon;

The input chart data must be valid CSV:

CSV data is similar to L</Just data> above, except the first line is used as
"xkey,ykey1,ykey2,...". Example:

  # Can have comments in CSV input as well
  x,a,b
  2015-02-04 15:03,120,90
  2015-03-14,75,65
  2015-04,100,40

CSV input data require L<Text::CSV> to be installed.

=head2 Embedding

A paste can be embedded in other pages using the query param "embed". Examples:

=over 2

=item * L<http://p.thorsen.pm/mojopastedemo.txt>

Get the raw data.

=item * L<http://p.thorsen.pm/mojopastedemo?embed=text>

Show the paste without any margin/padding and no menu.

=item * L<http://p.thorsen.pm/mojopastedemo/chart?embed=graph>

Show only the graph data.

=item * L<http://p.thorsen.pm/mojopastedemo/chart?embed=graph,heading,description>

Show the graph data, heading and description, but no menus.

=back

=head1 DEMO

You can try mojopaste here: L<http://p.thorsen.pm>.

=head1 INSTALLATION

Install system wide with cpanm:

  $ cpanm --sudo App::mojopaste

Don't have cpanm installed?

  $ curl -L http://cpanmin.us | perl - --sudo App::mojopaste
  $ wget http://cpanmin.us -O - | perl - --sudo App::mojopaste

It is also possible to install L<App::mojopaste> using Docker. Check out
L<https://hub.docker.com/r/jhthorsen/mojopaste> for more information.

=head1 SYNOPSIS

=over 2

=item * Simple single process daemon

  $ mojopaste daemon --listen http://*:8080

=item * Save paste to custom dir

  $ PASTE_DIR=/path/to/paste/dir mojopaste daemon --listen http://*:8080

=item * Using the UNIX optimized, preforking hypnotoad web server

  $ MOJO_CONFIG=/path/to/mojopaste.conf hypnotoad $(which mojopaste)

Example mojopaste.conf:

  {
    brand_link    => "index",
    brand_logo    => "/images/logo.png",
    brand_text    => "Mojopaste",
    paste_dir     => "/path/to/paste/dir",
    enable_charts => 1, # default is 0
    hypnotoad     => {
      listen => ["http://*:8080"],
    },
  }

"enable_charts" is for adding a button which can make a chart of the input
data using L<morris.js|http://morrisjs.github.io/morris.js>

Check out L<Mojo::Server::Hypnotoad> for more hypnotoad options.

=back

=head1 OTHER PASTEBINS

=over 2

=item * L<http://paste.scsys.co.uk>

=item * L<http://paste.ubuntu.com>

=item * L<http://pastebin.com>

=back

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut
